modules/mtl/arrow-mtl-data/src/test/kotlin/arrow/mtl/AccumTTest.kt

```diff
@@ -0,0 +1,203 @@
+import arrow.fx.fix
+import arrow.fx.mtl.accumt.monadIO.monadIO
+import arrow.mtl.extensions.accumt.alternative.alternative
+import arrow.mtl.extensions.accumt.applicative.applicative
+import arrow.mtl.extensions.accumt.applicativeError.applicativeError
+import arrow.mtl.extensions.accumt.functor.functor
+import arrow.mtl.extensions.accumt.monad.monad
+import arrow.mtl.extensions.accumt.monadError.monadError
+import arrow.mtl.extensions.accumt.monadTrans.monadTrans
+import arrow.test.UnitSpec
+import arrow.test.generators.GenK
+import arrow.test.generators.genK
+import arrow.test.generators.intSmall
+import arrow.test.generators.tuple2
+import arrow.test.laws.AlternativeLaws
+import arrow.test.laws.ApplicativeErrorLaws
+import arrow.test.laws.ApplicativeLaws
+import arrow.test.laws.FunctorLaws
+import arrow.test.laws.MonadErrorLaws
+import arrow.test.laws.MonadLaws
+import arrow.test.laws.MonadTransLaws
+import arrow.test.laws.equalUnderTheLaw
+import arrow.typeclasses.Eq
+import arrow.typeclasses.EqK
+import arrow.typeclasses.Monad
+import arrow.typeclasses.Monoid
+import io.kotlintest.properties.Gen
+import io.kotlintest.properties.forAll
+import io.kotlintest.shouldBe
+
+class AccumTTest : UnitSpec() {
+  init {
+    testLaws(
+      FunctorLaws.laws(
+        AccumT.functor<Int, ForId>(Id.functor()),
+        AccumT.genK(Id.genK(), Gen.intSmall()),
+        AccumT.eqK(Id.monad(), Id.eqK(), Int.eq(), 123)
+      ),
+
+      ApplicativeLaws.laws(
+        AccumT.applicative(Int.monoid(), Id.monad()),
+        AccumT.functor<Int, ForId>(Id.functor()),
+        AccumT.genK(Id.genK(), Gen.intSmall()),
+        AccumT.eqK(Id.monad(), Id.eqK(), Int.eq(), 123)
+      ),
+
+      MonadLaws.laws(
+        AccumT.monad(Int.monoid(), Id.monad()),
+        AccumT.genK(Id.genK(), Gen.intSmall()),
+        AccumT.eqK(Id.monad(), Id.eqK(), Int.eq(), 4711)
+      ),
+
+      MonadTransLaws.laws(
+        AccumT.monadTrans(String.monoid()),
+        Id.monad(),
+        AccumT.monad(String.monoid(), Id.monad()),
+        Id.genK(),
+        AccumT.eqK(Id.monad(), Id.eqK(), String.eq(), "hello")
+      ),
+
+      AlternativeLaws.laws<AccumTPartialOf<Int, ForOption>>(
+        AccumT.alternative(Option.alternative(), Option.monad(), Int.monoid()),
+        AccumT.genK(Option.genK(), Gen.int()),
+        AccumT.eqK(Option.monad(), Option.eqK(), Int.eq(), 10)
+      ),
+
+      ApplicativeErrorLaws.laws(
+        AccumT.applicativeError(Either.monadError<String>(), Int.monoid()),
+        AccumT.genK(Either.genK(Gen.string()), Gen.int()),
+        Gen.string(),
+        AccumT.eqK(Either.monad(), Either.eqK(String.eq()), Int.eq(), 10)
+      ),
+
+      MonadErrorLaws.laws<AccumTPartialOf<Int, EitherPartialOf<String>>, String>(
+        AccumT.monadError(Int.monoid(), Either.monadError()),
+        AccumT.genK(Either.genK(Gen.string()), Gen.int()),
+        Gen.string(),
+        AccumT.eqK(Either.monad(), Either.eqK(String.eq()), Int.eq(), 10)
+      )
```

Comment:
Each of the Laws already includes the ones from the previous typeclasses in the hierarchy. In this case just by testing MonadError you already test the others.

Reply:
:point_up: Also please add the `functor` and `applicative` instance to that test (there should be an overload taking those together with the `Monad` instance, just use `Monad` directly for those that are missing). This then tests consistency between the implementations and the default implementations in `Monad`